### DISCUSSION TITLE
iTwin V3 App does not show timeline

### DISCUSSION THREAD

**Discussion by oussamabonnor1** [iTwin/itwinjs-core#3559] [Q&A] [answered]
After migrating to the V3 app, timeline data is no longer being shown.
I tried loggin values of the schedule into console and got `undefined`

```typescript
viewPort.view.scheduleScript => undefined
viewPort.displayStyle.scheduleScript => undefined
```

Here are the dependencies I added on `package.json`

```json
"dependencies": {
    "@bentley/icons-generic": "^1.0.34",
    "@bentley/react-scripts": "^4.0.7",
    "@itwin/appui-abstract": "^3.1.3",
    "@itwin/appui-layout-react": "^3.1.3",
    "@itwin/appui-react": "^3.1.3",
    "@itwin/browser-authorization": "^0.5.1",
    "@itwin/components-react": "^3.1.3",
    "@itwin/core-bentley": "^3.1.3",
    "@itwin/core-common": "^3.1.3",
    "@itwin/core-frontend": "^3.1.3",
    "@itwin/core-geometry": "^3.1.3",
    "@itwin/core-i18n": "^3.1.3",
    "@itwin/core-markup": "^3.1.3",
    "@itwin/core-orbitgt": "^3.1.3",
    "@itwin/core-quantity": "^3.1.3",
    "@itwin/core-react": "^3.1.3",
    "@itwin/core-telemetry": "^3.1.3",
    "@itwin/imodel-components-react": "^3.1.3",
    "@itwin/imodels-access-frontend": "^1.0.1",
    "@itwin/imodels-client-management": "^1.0.1",
    "@itwin/itwinui-react": "^1.36.0",
    "@itwin/presentation-common": "^3.1.3",
    "@itwin/presentation-frontend": "^3.1.3",
    "@itwin/reality-data-client": "^0.7.2",
    "@itwin/web-viewer-react": "^2.0.9",
    "@itwin/webgl-compatibility": "^3.1.3",
    "@types/history": "^4.7.11",
    "@types/node": "^14.18.13",
    "@types/react": "^17.0.44",
    "@types/react-dom": "^17.0.15",
    "history": "^4.10.1",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-redux": "^7.2.8",
    "typescript": "~4.3.0"
  },
  "devDependencies": {
    "@bentley/react-scripts": "^4.0.3",
    "@types/history": "^4.7.6",
    "@types/node": "^14.0.19",
    "@types/react": "^17.0.9",
    "@types/react-dom": "^17.0.9",
    "typescript": "~4.3.0"
  },
```

P.s: This was tested with both normal sign in auth and server auth. Also, the iModels with timelines still show correctly in V2 (problem not in the iModel).

**Comment by pmconne** [comment]
Please see [this answer](https://github.com/iTwin/itwinjs-core/discussions/3006#discussioncomment-1990753).

**Reply by oussamabonnor1** [reply]
I'm sorry, I can't find any documentation that points where the `ContentProp` config should be affected.

```
const thisContentProps: ContentProps = {
        id: "imodel-view-default",
        classId: IModelViewportControl,
        applicationData:
        {
          viewState, iModelConnection,
          featureOptions:
          {
            defaultViewOverlay: {
              enableScheduleAnimationViewOverlay: true,
              enableAnalysisTimelineViewOverlay: true,
              enableSolarTimelineViewOverlay: true,
            },
          },
        },
      };
```

Could you tell me what's the quickest way to update a default app's settings to show the schedule ? After a bit of reading, I understood that the default frontstage of the app should be updated to reflect those changes, how can one do that ?

**Reply by pmconne** [reply]
I can't, but @NancyMcCallB or @bsteinbk should be able to.

**Reply by NancyMcCallB** [reply]
The featureOptions listed above are part of the iModelViewportOptions passed into the viewer in the viewportOptions argument.

Something like (I'm typing in place, so copy/paste with care):

const myFeatureOptions = {defaultViewOverlay: {enableScheduleAnimationViewOverlay: true}};
const viewportOptions = { featureOptions: myFeatureOptions};

<Viewer
/* other view args *
viewportOptions={viewportOptions}
/>

Here's some documentation about [ IModelViewportControl: ](https://www.itwinjs.org/learning/ui/appui-react/imodelviewportcontrol/)

I'll try to add a sandbox sometime today.

**Reply by oussamabonnor1** [reply]
Thank you, I'm waiting for that sandbox impatiently.

Meanwhile it seems adding this does not work.

```typescript
viewportOptions={{
                featureOptions: {
                  defaultViewOverlay: {
                    enableScheduleAnimationViewOverlay: true,
                  },
                }
              }}
```

I'm not sure what I'm missing. Is there anything I can try meanwhile a proper demo is published ?

**Comment by NancyMcCallB** [comment]
I can't seem to get this tested in a sandbox, but I have tried the code above in a Viewer and it works properly.

This code explicitly shows the schedule animation timeline. There are two others.

Which are you trying to display?

**Reply by bsteinbk** [reply]
Please also review the view properties ... if they are still "undefined" as you have shown, no timeline overlay would be displayed. The schedule script must be defined.

**Reply by oussamabonnor1** [reply]
> I can't seem to get this tested in a sandbox, but I have tried the code above in a Viewer and it works properly.
> 
> This code explicitly shows the schedule animation timeline. There are two others.
> 
> Which are you trying to display?

* I'm trying trying to only show the schedule animation timeline, I have no use for the other two for now.
-------------

> Please also review the view properties ... if they are still "undefined" as you have shown, no timeline overlay would be displayed. The schedule script must be defined.

They're `undefined` still, yes: 

I'm running an instance of `V2` with an iModel that contains a `timeline` ==> schedule script `defined`+ timeline `visible`.
Running a `V3` instance with the **same** iModel that contains a `timeline` ==> schedule script `undefined`+ timeline `invisible`.

Out of fear that my app was faulty, I created a new V3 test Viewer and it doesn't work as well.

Did anything change from V2 to V3 on how to setup timelines ? Mine are created in Synchro. Did the pipeline change in any way ? (previously a timeline was understood by iTwin without any extra steps and a schedule script was auto created/defined).

**Comment by oussamabonnor1** [comment]
I don't mean to be pushy but, any update on this, fellas ?

**Comment by pmconne** [comment, marked as answer]
Try logging `JSON.stringify(viewport.displayStyle.settings._json)`. If you're comfortable doing so, paste the output here. Otherwise, confirm whether it contains a `scheduleScript` and/or `renderTimeline` property.

**Reply by oussamabonnor1** [reply]
Sure, no scheduleScript or renderTimeline were found, although the same iModel loaded in the V2 app works perfectly.

```json
{
    "viewflags": {
        "renderMode": 6,
        "noSourceLights": false,
        "noCameraLights": false,
        "noSolarLight": false,
        "noConstruct": true,
        "noTransp": false,
        "visEdges": false,
        "backgroundMap": true
    },
    "mapImagery": {
        "backgroundBase": {
            "name": "Bing Maps: Aerial Imagery with labels",
            "visible": true,
            "transparentBackground": false,
            "url": "https://dev.virtualearth.net/REST/v1/Imagery/Metadata/AerialWithLabels?o=json&incl=ImageryProviders&key={bingKey}",
            "formatId": "BingMaps",
            "provider": {
                "name": "BingProvider",
                "type": 3
            }
        }
    },
    "environment": {
        "sky": {
            "skyColor": 16764303,
            "groundColor": 8244623,
            "nadirColor": 32040,
            "zenithColor": 16741686,
            "display": true
        },
        "ground": {
            "elevation": -0.01,
            "aboveColor": 25600,
            "belowColor": 2179941,
            "display": false
        }
    },
    "backgroundMap": {
        "applyTerrain": true,
        "terrainSettings": {
            "heightOriginMode": 0,
            "heightOrigin": 10
        }
    },
    "backgroundColor": 4294967295
}
```

**Reply by NancyMcCallB** [reply]
It is the same iModel, but is it the same view?

**Reply by oussamabonnor1** [reply]
I'm not sure if I understood correctly the meaning of `same view`, I'm running 2 separate apps, one iTwin V2 and the other iTwin V3, how could they have the same view ? If you mean a serialized view then no, I'm not saving/loading any.

As for the same result from the V2 app, a `scheduleScript` is there, same iModel.

```json
{
    "viewFlags": {
        "renderMode": 6,
        "dimensions": true,
        "patterns": true,
        "weights": true,
        "styles": true,
        "transparency": true,
        "fill": true,
        "textures": true,
        "materials": true,
        "acsTriad": false,
        "grid": false,
        "visibleEdges": false,
        "hiddenEdges": false,
        "sourceLights": false,
        "cameraLights": true,
        "solarLight": false,
        "shadows": false,
        "clipVolume": true,
        "constructions": false,
        "monochrome": false,
        "noGeometryMap": false,
        "backgroundMap": false,
        "hLineMaterialColors": false,
        "edgeMask": 0,
        "ambientOcclusion": false,
        "thematicDisplay": false,
        "forceSurfaceDiscard": false,
        "whiteOnWhiteReversal": true
    },
    "backgroundColor": 16777215,
    "scheduleScript": [
        {
            "modelId": "0x20000000003",
            "elementTimelines": [
                ...
            ]
        }
    ],
    "viewflags": {
        "renderMode": 6,
        "noSourceLights": false,
        "noCameraLights": false,
        "noSolarLight": false,
        "noConstruct": true,
        "noTransp": false,
        "visEdges": false,
        "backgroundMap": true
    },
    "timePoint": 1645030800,
    "environment": {
        "sky": {
            "display": true,
            "skyColor": 16764303
        }
    }
}
```

**Reply by aruniverse** [reply]
Can you check the changesetId is the same across V2 and V3? You can get that off of the iModelConnection

**Reply by pmconne** [reply]
@oussamabonnor1 So, you are allowing the Viewer to create a view for you, rather than loading a specific persistent view.
The Viewer used to merge in some properties of the iModel's "default" view, which could include the schedule script.
This was considered "bad". @kckst8, the Viewer was updated to no longer do that, correct? If so that will explain the difference.

**Reply by oussamabonnor1** [reply]
> Can you check the changesetId is the same across V2 and V3? You can get that off of the iModelConnection
Both the V2/V3 apps are loading the same changeset:

```json
{
    "id": "3018d857bed925b2fea02207ec421c711504008d",
    "index": 9
}
```

**Reply by oussamabonnor1** [reply]
> @oussamabonnor1 So, you are allowing the Viewer to create a view for you, rather than loading a specific persistent view. The Viewer used to merge in some properties of the iModel's "default" view, which could include the schedule script. This was considered "bad". @kckst8, the Viewer was updated to no longer do that, correct? If so that will explain the difference.

-------
What's the best practice to load a specific persistent view ? How to create one initially ? I'd love to get a link to try that approach.

About the Viewer merging properties, every time we save a [Serialize View](https://www.itwinjs.org/sandboxes/iTwinPlatform/Serialize%20View) in an iModel that contains a schedule, the properties get merged into a single weird bundle and the result is very similar to what I keep getting with the V3 based app (thousands of lines of detailed schedule data embedded but no longer understood by iTwin). Maybe that info could help.

**Reply by pmconne** [reply]
@kckst8 please show @oussamabonnor1 how to configure the viewer to load a persistent view.
@oussamabonnor1 iModels produced by SYNCHRO generally include a default persistent view called something like "Schedule View", that's the one that would host the schedule script.

> the result is very similar to what I keep getting with the V3 based app (thousands of lines of detailed schedule data embedded but no longer understood by iTwin)

Can you clarify please? You said V3 gets an `undefined` scheduleScript, not thousands of lines of schedule data.

**Reply by oussamabonnor1** [reply]
> @kckst8 please show @oussamabonnor1 how to configure the viewer to load a persistent view.

If there's a sample or a sandbox example that I can use, please let me know. I'd rather not bother too many people :)

> Can you clarify please? You said V3 gets an undefined scheduleScript, not thousands of lines of schedule data.

Yes, sorry. I should've been more clear, the V2 app was the one to produce that unusual result.

Here's an example of what was given after saving https://jsfiddle.net/t4jfe9us/

**Reply by kckst8** [reply]
Hi @oussamabonnor1 . Most of the sandboxes do this (load a persisted view state): https://www.itwinjs.org/sandboxes/JoshSchifter/facet-solids-2.x. There is a viewState property in the viewportOptions prop that accepts a full viewState to override the default that the Viewer provides. If you need to fetch the viewState asynchronously, you can do so in the onIModelConnected callback, as is done in the sandbox referenced above.

**Reply by oussamabonnor1** [reply]
Thanks, this is uncharted territory for me so I have a few questions if I may:

* What is the goal of overriding the default viewState provided by the Viewer.
* All samples use `ViewSetup` pulled from `@itwin/sandbox`, is this something available for us or is there another way to get access to the `Default View`.
* The `ViewSetup` provides a `Default View` that acts as a viewState, does it mean it was created by default by iTwin (because then, why use it at all) ?
* How will providing a default viewState manually help resolve the schedule issue ?

Excuse any silly questions, I am fairly new to this concept.

**Reply by NancyMcCallB** [reply]
Hi, @oussamabonnor1 ,
I think some of what you need to do may be obscured by the @itwin/sandbox. I found that it's easiest to make a simple frontstage and pass it in to the Viewer as a new default stage. I've added the code to this sandbox: https://www.itwinjs.org/sandboxes/NancyMcCall/Default%20View%20Frontstage

In it, I get the default view id by querying the  IModelConnection and then I use that to get the default view's viewstate.

The frontstage itself is quite basic and registers the standard tool widgets and status bar.  Let me  know if this helps you get to your schedule timeline and whether you have more questions.

**Reply by oussamabonnor1** [reply]
Thank you @NancyMcCallB! This seems to have done the trick.

Now off to redesign/reassign every UI related update I've done so far to the new frontstage 😄 

Thank you @kckst8, @pmconne, @aruniverse & @bsteinbk. I deeply appreciate your time and effort 😃

**Reply by jptissot** [reply]
Hi, @NancyMcCallB I ran into the same issue and would like to keep the default FrontStage but conditionally load the default view. Is there an easy way to do this without creating a new FrontStage?

**Reply by NancyMcCallB** [reply]
Hi, @jptissot ,

You should be able to use the logic from the provideConentGroup() that sets the viewState in your own code to set up the view (that's basically what ViewSetup does). You'll want to wait for the iModelConnection to be established, though. If you run into trouble, let me know and I'll try it out in a ViewerApp.

Thanks,
Nancy
